app/models/theft_alert.rb

```diff
@@ -0,0 +1,79 @@
+class TheftAlert < ActiveRecord::Base
+  enum status: { pending: 0, active: 1, inactive: 2 }.freeze
+
+  validates :stolen_record,
+            :theft_alert_plan,
+            :status,
+            :creator,
+            presence: true
+
+  belongs_to :stolen_record
+  belongs_to :theft_alert_plan
+  belongs_to :payment
+  belongs_to :creator,
+             class_name: "User",
+             foreign_key: :user_id
+
+  scope :due_for_expiration, -> { active.where('"theft_alerts"."end_at" <= ?', Time.current) }
+
+  delegate :bike, to: :stolen_record
+
+  def begin!(kwargs)
+    start_time = Time.current.beginning_of_day
+    end_time = start_time + theft_alert_plan&.duration_days&.days
+
+    attrs = {
+      status: "active",
+      facebook_post_url: kwargs.fetch(:facebook_post_url, ""),
+      begin_at: start_time,
+      end_at: end_time.end_of_day,
+    }
+
+    update(attrs) if valid_state?(attrs)
+  end
+
+  def end!(*)
+    attrs = {
+      status: "inactive",
+      facebook_post_url: facebook_post_url,
+      begin_at: begin_at,
+      end_at: end_at,
+    }
+
+    update(attrs) if valid_state?(attrs)
+  end
+
+  def reset!(*)
+    attrs = {
+      status: "pending",
+      facebook_post_url: "",
+      begin_at: nil,
+      end_at: nil,
+    }
+
+    update(attrs) if valid_state?(attrs)
+  end
+
+  private
+
+  # Ensure fields have expected values for the target state's status
+  def valid_state?(target_state)
```

Comment:
I think I'd prefer this to either:

Rather than passing a hash of target state, assign the attrs to the `TheftAlert` and use a [custom validator](https://guides.rubyonrails.org/active_record_validations.html#custom-methods) like `validate :record_in_valid_state` - which will block save if in a state is invalid. This has the advantage of following rails conventions, and enabling things like re-rendering a form with the values that were input, while displaying the error.

Reply:
I'd prefer the same, but I ran into some friction with that approach. I'll circle back to trace through the execution order and refactor.

Reply:
If you want to deal with refactoring it later that's fine.